genesis/engine/couplers/sap_coupler.py

```diff
@@ -2781,22 +3075,24 @@ def detection(self, f: ti.i32):
 
     @ti.func
     def compute_delassus_world_frame(self):
-        dt2 = self.sim._substep_dt**2
+        dt2_inv = 1.0 / (self.sim._substep_dt**2)
         # rigid
-        self.coupler.rigid_solve_contact(
-            self.Jt, self.M_inv_Jt, self.n_contact_pairs[None], self.contact_pairs.batch_idx
+        self.coupler.rigid_solve_jacobian(
+            self.Jt, self.M_inv_Jt, self.n_contact_pairs[None], self.contact_pairs.batch_idx, 3
         )
         self.W.fill(0.0)
         for i_p, i_d, i, j in ti.ndrange(self.n_contact_pairs[None], self.rigid_solver.n_dofs, 3, 3):
-            self.W[i_p][i, j] += self.M_inv_Jt[i_p, i_d][i] * self.Jt[i_p, i_d][j] * dt2
+            self.W[i_p][i, j] += self.M_inv_Jt[i_p, i_d][i] * self.Jt[i_p, i_d][j]
 
         # fem
         for i_p in range(self.n_contact_pairs[None]):
             i_g0 = self.contact_pairs[i_p].geom_idx0
             i_b = self.contact_pairs[i_p].batch_idx
             for i in ti.static(range(4)):
                 i_v = self.fem_solver.elements_i[i_g0].el2v[i]
-                self.W[i_p] += self.contact_pairs[i_p].barycentric0[i] ** 2 * self.fem_solver.pcg_state_v[i_b, i_v].prec
+                self.W[i_p] += (self.contact_pairs[i_p].barycentric0[i] ** 2 * dt2_inv) * self.fem_solver.pcg_state_v[
```

Comment:
You don't need parenthesis here. It may even prevent optimisation from the compiler (at least when it is running on CPU)

Besides, it could be great to create a temporary for `self.contact_pairs[i_p].barycentric0` to avoid split on multiple lines, which is a bit hard to read.